### PR TITLE
oast: persist alert details in permanent DB

### PR DESCRIPTION
**Pull request by thc202**
Keep the alert details (and message) to be able to raise the alerts when no longer using the original session.

### CODE REVIEW

File: addOns/oast/src/main/java/org/zaproxy/addon/oast/internal/AlertEntity.java
Comment (kingthorin) [inline addOns/oast/src/main/java/org/zaproxy/addon/oast/internal/AlertEntity.java:1]: Should this be persisting source? Or is that derived from the message?

File: addOns/oast/src/main/java/org/zaproxy/addon/oast/internal/OastPermanentDatabase.java [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+            var entity = query.setParameters(payload).executeUnique();
+            pm.retrieve(entity);
+            pm.retrieve(entity.getMessage());
+            return entity;
+        } catch (JDOObjectNotFoundException e) {
+            return null;
+        } catch (InterruptedException e) {
+            Thread.currentThread().interrupt();
+        } catch (Exception e) {
+            LOGGER.error("An error occurred while getting the alert:", e);
```
Comment (kingthorin) [inline addOns/oast/src/main/java/org/zaproxy/addon/oast/internal/OastPermanentDatabase.java:50]: Space after the colon?

File: addOns/oast/src/main/java/org/zaproxy/addon/oast/internal/MessageEntity.java [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+import javax.jdo.annotations.Column;
+import javax.jdo.annotations.IdGeneratorStrategy;
+import javax.jdo.annotations.PersistenceCapable;
+import javax.jdo.annotations.Persistent;
+import javax.jdo.annotations.PrimaryKey;
+import org.datanucleus.api.jdo.annotations.CreateTimestamp;
+import org.parosproxy.paros.network.HttpBody;
+import org.parosproxy.paros.network.HttpMessage;
+
+@Cacheable("false")
```
Comment (kingthorin) [inline addOns/oast/src/main/java/org/zaproxy/addon/oast/internal/MessageEntity.java:33]: I noticed this on both things. It's probably a different cache but more so I understand. Cache is off here but the SQL creates it as a cached table?

File: addOns/oast/src/main/java/org/zaproxy/addon/oast/internal/OastPermanentDatabase.java (hunk 1 above)
Comment (thc202) [reply addOns/oast/src/main/java/org/zaproxy/addon/oast/internal/OastPermanentDatabase.java]: Done.

File: addOns/oast/src/main/java/org/zaproxy/addon/oast/internal/AlertEntity.java
Comment (thc202) [reply addOns/oast/src/main/java/org/zaproxy/addon/oast/internal/AlertEntity.java:1]: Source is being persisted.

File: addOns/oast/src/main/java/org/zaproxy/addon/oast/internal/MessageEntity.java (hunk 2 above)
Comment (thc202) [reply addOns/oast/src/main/java/org/zaproxy/addon/oast/internal/MessageEntity.java:32]: This is for L2 cache, which is not the same as the underlying database.

**Review by psiinon — APPROVED** [review]

**Comment by psiinon** [comment]
WFM 😁

File: addOns/oast/src/main/java/org/zaproxy/addon/oast/internal/AlertEntity.java
Comment (kingthorin) [reply addOns/oast/src/main/java/org/zaproxy/addon/oast/internal/AlertEntity.java:1]: I must have scrolled right by it, sorry

File: addOns/oast/src/main/java/org/zaproxy/addon/oast/internal/AlertEntity.java [hunk 3]
```diff
@@ -0,0 +1,316 @@
...
+        param = alert.getParam();
+        attack = alert.getAttack();
+        otherInfo = alert.getOtherInfo();
+        solution = alert.getSolution();
+        reference = alert.getReference();
+        evidence = alert.getEvidence();
+        inputVector = alert.getInputVector();
+        cweId = alert.getCweId();
+        wascId = alert.getWascId();
+        sourceId = alert.getSource() != null ? alert.getSource().getId() : 0;
```
Comment (kingthorin) [inline addOns/oast/src/main/java/org/zaproxy/addon/oast/internal/AlertEntity.java:114]: Could you switch this to use positive logic?

I know it'll be the same result but it'll be easier to read for future maintenance

**Review by kingthorin — COMMENTED** [review]
Not necessarily for this PR but as we use this DB stack more is there some sort of config file we can use that will apply both to the Java entities and sql so we don't have to define/maintain column sizes in two places?

File: addOns/oast/src/main/java/org/zaproxy/addon/oast/internal/AlertEntity.java (hunk 3 above)
Comment (thc202) [reply addOns/oast/src/main/java/org/zaproxy/addon/oast/internal/AlertEntity.java:114]: What do you mean with positive? `== null`?

**Comment by thc202** [comment]
https://github.com/zaproxy/zap-extensions/pull/5710#pullrequestreview-2282610269

Technically they are not maintained in two places, what's in the script is what matters but the definition in the code helps datanucleus know what the limit is.

File: addOns/oast/src/main/java/org/zaproxy/addon/oast/internal/AlertEntity.java (hunk 3 above)
Comment (thc202) [reply addOns/oast/src/main/java/org/zaproxy/addon/oast/internal/AlertEntity.java]: Changed.

Comment (kingthorin) [reply addOns/oast/src/main/java/org/zaproxy/addon/oast/internal/AlertEntity.java]: Yes, that's what I was after.

**Review by kingthorin — APPROVED** [review]
LGTM

**Comment by kingthorin** [comment]
Merging, was tested live by @psiinon

**Pull request merged**